### PR TITLE
Add map_first_graphql_response, unify execution & supergraph response type

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/apollographql/router/issues/1564, see discussion there.

### CODE REVIEW

**Review by garypen — APPROVED** [review]
Looks good. Some documentation nits.

File: NEXT_CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,46 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
...
+### New plugin helper: `map_first_graphql_response`
+
+In supergraph and execution services, the service response contains
+not just one GraphQL response but a stream of them,
+in order to support features such as `@defer`.
+
+This new method of `ServiceExt` and `ServiceBuilderExt` in `apollo_router::layers`
+wraps a service and call `callback` when the first GraphQL response
+in the stream returned by the inner service becomes available.
+The callback can then modify the HTTP parts (headers, status code, etc)
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply NEXT_CHANGELOG.md:45]: `http::response::Parts` is also in a the return tuple, so the callback can mutate it before returning. This ability is why the first GraphQL response is special, as we have it when headers aren’t sent yet

[the remark this answers is not in the document]
Comment (garypen) [reply NEXT_CHANGELOG.md:45]: I know. I think it's a language thing. A user may want to "access" parts without "modifying" them. I was trying to think of a verb that covered both read and write access. I'm fine with leaving as is if that's clear enough.

**Review by o0Ignition0o — APPROVED** [review]
Love it!

**Pull request merged**